File: Random/LC_13.java
// https://leetcode.com/problems/roman-to-integer/

public class LC_13 {
    public static int romanToInt(String s) {
        // res is our output, num is the current character in number form
        // prev is the previous number
        int res = 0, num = 0, prev = 0;

        for (int i = s.length() - 1; i >= 0; i--) {
            switch (s.charAt(i)) {
                // We do a normal mapping with the roman literals with integer values for future access
                case 'I' -> num = 1;
                case 'V' -> num = 5;
                case 'X' -> num = 10;
                case 'L' -> num = 50;
                case 'C' -> num = 100;
                case 'D' -> num = 500;
                case 'M' -> num = 1000;
            }

            // If the previous number was larger than the current one, we subtract the current number from the result
            // This is for cases like CM, IV
            if (prev > num) {
                res -= num;
            } else {
                // Otherwise just add it in the result
                res += num;
            }

            // We add the current number in previous since for the next iteration it will be the previous
            prev = num;
        }

        return res;
    }

    public static void main(String[] args) {
        System.out.println(romanToInt("MCMXCIV"));
    }
}

File: LeetCode JS/LC_13.js
// https://leetcode.com/problems/roman-to-integer/

/**
 * Converts a Roman numeral string to an integer value.
 *
 * @param {string} s - The Roman numeral string to be converted.
 * @return {number} - The integer value of the Roman numeral.
 */
const romanToInt = (s) => {
	let prev = 0,
		res = 0
	for (let i = s.length - 1; i >= 0; i--) {
		let n = getNumber(s[i])

		if (prev > n) {
			res -= n
		} else {
			res += n
		}

		prev = n
	}

	return res
}

/**
 * Returns the corresponding number for a given Roman numeral.
 *
 * @param {string} ch - The Roman numeral character to convert.
 * @return {number} The corresponding decimal number.
 */
const getNumber = (ch) => {
	switch (ch) {
		case 'I':
			return 1
		case 'V':
			return 5
		case 'X':
			return 10
		case 'L':
			return 50
		case 'C':
			return 100
		case 'D':
			return 500
		case 'M':
			return 1000
	}
}

console.log(romanToInt('III'))
